### PR TITLE
584 worker decommissioning

### LINKED ISSUE #584: HDFS: Support worker decommissioning
TDP should support decommissioning worker nodes. Specific actions have to be performed for HDFS DataNode and YARN ResourceManager components.

See [HDP - Decommissioning slave nodes](https://docs.cloudera.com/HDPDocuments/HDP3/HDP-3.1.4/administration/content/decommissioning-slave-nodes.html) for reference.

### PR DESCRIPTION
**Pull request by Nuttymoon**
#### Which issue(s) this PR fixes

Fixes #584 

#### Additional comments

This PR adds the possibility of easily decommissioning workers in TDP clusters. I created 2 variables to allow this: `hdfs_datanodes_decommission` and `yarn_nodemanagers_decommission`.

Procedure to follow:
1. Add the FQDNs of the workers to decommission to `hdfs_datanodes_decommission` and `yarn_nodemanagers_decommission`
   ```yaml
   hdfs_datanodes_decommission:
     - worker-04.tdp
   
   yarn_nodemanagers_decommission:
     - worker-04.tdp
   ```
2. Deploy the configuration (e.g. `tdp reconfigure`)
3. Refresh the workers' lists:
    - In HDFS: `hdfs dfsadmin -refreshNodes`
    - In YARN: `yarn rmadmin -refreshNodes`
4. Wait for the decommissioning to finish (the easiest way is from the web UIs)
5. Shutdown the DataNode and NodeManager services on the worker (or shutdown the machine)
6. (Optional) Remove the worker FQDN from the `decommission` lists

Maybe we can create a playbook that unrolls this procedure but from what I know there is no playbook in TDP that perform maintenance operations and I don't know if we want to have some.

**Note:** The implementation takes strong inspiration from [HDP - Decommissioning slave nodes](https://docs.cloudera.com/HDPDocuments/HDP3/HDP-3.1.4/administration/content/decommissioning-slave-nodes.html)

#### Agreements

- [x] I hereby declare this contribution to be licensed under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0).
- [x] I hereby agree to grant [TOSIT](https://www.tosit.io/) a copyright license to use my contributions.

### CODE REVIEW

File: roles/hdfs/common/templates/dfs.exclude.j2 [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{% for dn in hdfs_datanodes_decommission %}
+{{ dn }}
+{% endfor %}
```
Comment (gboutry) [inline roles/hdfs/common/templates/dfs.exclude.j2:3]:
```suggestion
{% endfor %}

```

File: roles/yarn/common/templates/yarn.exclude.j2 [hunk 2]
```diff
@@ -0,0 +1,3 @@
+{% for nm in yarn_nodemanagers_decommission %}
+{{ nm }}
+{% endfor %}
```
Comment (gboutry) [inline roles/yarn/common/templates/yarn.exclude.j2:3]:
```suggestion
{% endfor %}

```

File: roles/hdfs/common/templates/dfs.exclude.j2 (hunk 1 above)
Comment (leopaul36) [inline roles/hdfs/common/templates/dfs.exclude.j2:3]: Missing new line at EOF

File: roles/yarn/common/templates/yarn.exclude.j2 (hunk 2 above)
Comment (leopaul36) [inline roles/yarn/common/templates/yarn.exclude.j2:3]: Missing new line at EOF

**Review by gboutry — CHANGES_REQUESTED** [review]

**Review by leopaul36 — APPROVED** [review]

**Pull request merged**